File: GFG/270125/Java/LRUCache-DSDesign.java
// Problem Link: https://www.geeksforgeeks.org/problems/lru-cache/1

// Approach 1: With LinkedHashMap
class LRUCache {
    // Constructor for initializing the cache capacity with the given value.
    
    /**
     * 
     *  Why LinkedHashMap?
     * LinkedHashMap instead of a normal HashMap in an LRU (Least Recently Used) cache is that LinkedHashMap maintains access order, 
     * which is crucial for efficiently identifying and removing the least recently used entry.
     * 
     * 
         In a normal HashMap, 
         you cannot directly remove the eldest entry when the capacity exceeds
     */
    static LinkedHashMap<Integer, Integer> map ;
    private final int capacity;
    
    LRUCache(int cap) {
        // code here
        this.capacity = cap;
        
        // The method checks if the size of the map exceeds the defined capacity (capacity).
        // If it does, the eldest (least recently used) entry is removed from the cache. 
        map = new LinkedHashMap<Integer, Integer>(capacity, 0.75f, true){
            protected boolean removeEldestEntry(Map.Entry eldest) {
                return size() > capacity;
            }
        };
        
    }

    // Function to return value corresponding to the key.
    public static int get(int key) {
        // your code here
        return map.getOrDefault(key, -1);
    }

    // Function for storing key-value pair.
    public static void put(int key, int value) {
        // your code here
        
        map.put(key, value);
    }
}

// Approach 2: With Normal Hashmap and LinkedList
// https://takeuforward.org/data-structure/implement-lru-cache/

File: GFG/270125/Python/LRUCache-DSDesign.py
# Problem Link: https://www.geeksforgeeks.org/problems/lru-cache/1

from collections import OrderedDict
class LRUCache:
      
    #Constructor for initializing the cache capacity with the given value.  
    def __init__(self, cap):
        #code here
        self.capacity =cap
        self.cache = OrderedDict()
        
        
    #Function to return value corresponding to the key.
    def get(self, key):
        #code here
        if key not in self.cache:
            return -1
        self.cache.move_to_end(key) # In LRU SYSTEM, WHENEVER YOU USE ONE ITEM, MOVE IT TO END (MOST RECENTLY USED)
        return self.cache[key]
        
        
    #Function for storing key-value pair.   
    def put(self, key, value):
        #code here
        if key in self.cache:
           self.cache.move_to_end(key)
           
           
        elif len(self.cache) >= self.capacity:
            self.cache.popitem(last= False)
        
        
        self.cache[key] = value
        self.cache.move_to_end(key)
